c/ckb_mmr.h

```diff
@@ -0,0 +1,463 @@
+#endif
+
+enum MMRErrorCode {
+  // MMR
+  ERROR_INVALID_STACK = 80,
+  ERROR_INVALID_COMMAND,
+  ERROR_INVALID_PROOF,
+  ERROR_PROOF_EOF,
+  ERROR_LEAF_EOF,
+  ERROR_NO_MORE_LEAFS,
+  ERROR_NO_MORE_COMMANDS,
+  ERROR_NODE_EOF,
+};
+
+#define MMR_NODE_TYPE_BUFFER 1
+#define MMR_NODE_TYPE_POINTER 2
+
+typedef struct {
+  uint8_t t;
+  union {
+    uint8_t buffer[MMR_NODE_BUFFER_MAX_BYTES];
+    const uint8_t *pointer;
+  } value;
+  uint32_t length;
+} mmr_node_t;
+
+typedef int (*mmr_leaf_reader_f)(void *, mmr_node_t *, uint64_t *);
+typedef int (*mmr_command_reader_f)(void *, uint8_t *);
+typedef int (*mmr_proof_reader_f)(void *, mmr_node_t *);
+
+// In the default case, one should pass the pointer to
+// mmr_default_buffer_reader_t data structure as reader contexts. A rough sample
+// code look as follows:
+//
+// ```
+// // First, load root_buffer, mmr_size, proof_buffer and leaf_buffer
+// // using syscalls.
+//
+// mmr_default_buffer_reader_t proof_buffer_reader;
+// mmr_default_buffer_reader_init(&proof_buffer_reader, proof_buffer,
+//                                proof_length);
+//
+// mmr_default_buffer_reader_t leaf_buffer_reader;
+// mmr_default_buffer_reader_init(&leaf_buffer_reader, leaf_buffer,
+//                                leaves_length);
+//
+// return mmr_verify(root_buffer, 32, mmr_size, &proof_buffer_reader,
+//                   &leaf_buffer_reader);
+// ```
+int mmr_verify(const uint8_t *root, uint32_t root_length, uint64_t mmr_size,
+               void *proof_reader_context, void *leaf_reader_context);
+
+const uint8_t *mmr_node_value(const mmr_node_t *node) {
+  if (node->t == MMR_NODE_TYPE_POINTER) {
+    return node->value.pointer;
+  }
+  return node->value.buffer;
+}
+
+// Note dst might collide with either lhs or rhs to save memcpy operation
+int mmr_blake2b_merge(mmr_node_t *dst, const mmr_node_t *lhs,
+                      const mmr_node_t *rhs) {
+  blake2b_state ctx;
+  ckb_blake2b_init(&ctx, 32);
+  blake2b_update(&ctx, mmr_node_value(lhs), lhs->length);
+  blake2b_update(&ctx, mmr_node_value(rhs), rhs->length);
+  blake2b_final(&ctx, dst->value.buffer, 32);
+  dst->t = MMR_NODE_TYPE_BUFFER;
+  dst->length = 32;
+  return 0;
+}
+
+typedef struct {
+  const uint8_t *buffer;
+  uint32_t buffer_length;
+  uint32_t index;
+} mmr_default_buffer_reader_t;
+
+int mmr_default_buffer_reader_init(mmr_default_buffer_reader_t *context,
+                                   const uint8_t *buffer,
+                                   uint32_t buffer_length) {
+  context->buffer = buffer;
+  context->buffer_length = buffer_length;
+  context->index = 0;
+  return 0;
+}
+
+int _mmr_default_buffer_read_node(void *context, mmr_node_t *out) {
+  mmr_default_buffer_reader_t *c = (mmr_default_buffer_reader_t *)context;
+
+  if (c->buffer_length - c->index < 2) {
+    return ERROR_NODE_EOF;
+  }
+  uint16_t len = *((uint16_t *)(&c->buffer[c->index]));
+  if (c->buffer_length - c->index < ((uint32_t ) 2) + len) {
```

Comment:
The value of `len` is extracted from the `proof` bytes, which are usually provided by the caller of a contract.

A potential attacker may encode `len` with a large number to cause an overflow in this line. Then the attacker can write unexpected content into the memory.